### PR TITLE
Submodule upgrades

### PR DESCRIPTION
**Pull request by dtronmans**
## Purpose
- Submodules commit hashes were old and are upgraded in this PR
- `ultralytics` package is now a required package for yolov5
- Match DetectV26, SegmentV26 and PoseV26 heads to match updated Ultralytics heads
## Summary by CodeRabbit

* **Chores**
  * Added `requests>=2.32.2` as a new dependency
  * Updated bundled YOLO submodule references
* **Refactor**
  * Improved end-to-end NMS-free box decoding for detection
  * Refined segmentation prototype and mask coefficient handling
  * Reworked pose keypoint decoding for more consistent decoding behavior

### CODE REVIEW

File: tools/modules/heads.py [hunk 1]
```diff
@@ -571,11 +572,8 @@ def forward(self, x):
         boxes = []
         scores = []
         for i in range(self.nl):
-            box = self.cv2[i](x[i])
-
-            cls_regress = self.cv3[i](x[i])
-            boxes.append(box.view(bs, 4, -1))
-            scores.append(cls_regress.view(bs, self.nc, -1))
+            boxes.append(self.cv2[i](x[i]).view(bs, 4 * self.reg_max, -1))
```
Comment (klemen1999) [inline tools/modules/heads.py]: As far as I can see this is just a refactor right? TBH I like more explicit code from before better with clearn box and cls_regress local variables

File: tools/modules/heads.py [hunk 2]
```diff
@@ -590,7 +588,6 @@ def forward(self, x):
 
     def _get_decode_boxes(self, preds):
         # Emulate ultralytics.nn.modules.head.Detect._get_decode_boxes for end2end export.
-        # preds["boxes"]: (N, 4, A), preds["feats"]: list of feature maps (N, C, H_i, W_i)
```
Comment (klemen1999) [inline tools/modules/heads.py:602]: I generally like comments like this in the code

File: tools/modules/heads.py [hunk 3]
```diff
@@ -562,6 +562,7 @@ def __init__(
         self.reg_max = old_detect.reg_max  # DFL channels
         self.no = old_detect.no  # number of outputs per anchor
         self.stride = old_detect.stride  # strides computed during build
+        self.dfl = old_detect.dfl
```
Comment (klemen1999) [inline tools/modules/heads.py]: This one is interesting. Wasn't DFL explicitly removed from YOLO26 architecture? IIRC this was one of the things that they explicitly mention in the paper/documentation

File: tools/modules/heads.py [hunk 4]
```diff
@@ -609,10 +606,8 @@ def _get_decode_boxes(self, preds):
             self.strides = stride_tensor.transpose(0, 1)
             self.shape = shape
 
-        # anchors: (1, 2, A), strides: (1, 1, A)
```
Comment (klemen1999) [inline tools/modules/heads.py:612]: I'd keep the comment

File: tools/modules/heads.py [hunk 5]
```diff
@@ -701,17 +696,9 @@ def forward(self, x):
...
-            boxes.append(box.view(bs, 4, -1))
-
-            # Class scores
-            cls_regress = self.cv3[i](x[i])
-            scores.append(cls_regress.view(bs, self.nc, -1))
-
-            # Mask coefficients
-            mask = self.cv4[i](x[i])
-            mask_coeffs.append(mask.view(bs, self.nm, -1))
+            boxes.append(self.cv2[i](x[i]).view(bs, 4 * self.reg_max, -1))
```
Comment (klemen1999) [inline tools/modules/heads.py]: same as before, more explicit code is I think in this case better

File: tools/modules/heads.py [hunk 6]
```diff
@@ -796,27 +783,17 @@ def forward(self, x):
...
-            # Box regression
-            box = self.cv2[i](x[i])
-            boxes.append(box.view(bs, 4, -1))
-
-            # Class scores
-            cls_regress = self.cv3[i](x[i])
-            scores.append(cls_regress.view(bs, self.nc, -1))
-
-            # Keypoints: cv4 extracts features, cv4_kpts predicts keypoints
+            boxes.append(self.cv2[i](x[i]).view(bs, 4 * self.reg_max, -1))
```
Comment (klemen1999) [inline tools/modules/heads.py]: same as before, more explicit code is I think in this case better

File: tools/modules/heads.py [hunk 7]
```diff
@@ -825,48 +802,27 @@ def forward(self, x):
...
-        # Note: After _get_decode_boxes, self.anchors is (2, A) and self.strides is (1, A)
         kpts_cat = torch.cat(kpts_raw, dim=2)  # (bs, nk, num_anchors)
-        kpts_decoded = self._kpts_decode(bs, kpts_cat)  # (bs, nk, num_anchors)
+        kpts_decoded = self._kpts_decode(kpts_cat)  # (bs, nk, num_anchors)
         kpts_decoded = kpts_decoded.permute(0, 2, 1)  # (bs, num_anchors, nk)
 
         return y, kpts_decoded
 
-    def _kpts_decode(self, bs, kpts):
+    def _kpts_decode(self, kpts):
```
Comment (klemen1999) [inline tools/modules/heads.py]: Do I understand correctly that this is an actual logic difference in how kpts are decoded right? 

---
Agent is saying that before and after are equivalent only if:
`kpts == y.view(bs, num_kpts, ndim, num_anchors)`
uses the same memory/layout interpretation as:
```
kpts[:, 0::ndim]  # all x channels
kpts[:, 1::ndim]  # all y channels
kpts[:, 2::ndim]  # visibility
```
That means kpts must be laid out as:
`[x0, y0, v0, x1, y1, v1, ...]`
per anchor.

---

Since (AFAIK) Ultralytics didn't make breaking changes to the model architecture I assume this is indeed true and these two are equivalent? Or did you see that with the old version the model wasn't working correctly?

Comment (klemen1999) [reply tools/modules/heads.py]: I'm fine with the change, just making sure that we don't need some fallback mechanism and this new implementation works correctly with all versions. But as I said, AFAIK there shouldn't be multiple YOLO26 model revisions so I suspect this is the case and we don't need any extra fallback

**Review by klemen1999 — APPROVED** [review]

File: tools/modules/heads.py [hunk 8]
```diff
@@ -744,7 +741,7 @@ def _get_proto(self, x):
 
         Proto26 takes all feature maps and returns prototype masks.
         """
-        return self.proto(x, return_semseg=False)
+        return self.proto(x, return_semantic=False)
```
[the remark this answers is not in the document]
Comment (klemen1999) [reply tools/modules/heads.py:744]: I don't think this is true. We use specific `ultralytics` commit hash (`d03e470ad14a5505c66985e27eb9e752c877983c`) here and in this specific one Proto26.forward() should still accept the `return_semantic` keyword argument. 
Double check though
